edx-olive/edx-platform
# [SE-4238] Periodic task execution of grade and profile reports

This PR adds a new Django app that allows running `calculate_grades_csv` and `calculate_students_features_csv` as periodic tasks configured on Django admin, described in Phase 1 in [SE-4238](https://tasks.opencraft.com/browse/SE-4238) until the platform version is not upgraded to Juniper or Koa.

**Dependencies**: This change depends on https://github.com/edx-olive/edx-platform/pull/51 because the generated report file names should contain the CCX ID if available. In case we decide that shouldn't block this PR, we can go ahead and merge to `pooja/cherry-pick-pymongo-upgrade` branch after removing the changes related to the mentioned PR.

**Screenshots**:

Example Django admin configuration for periodic build
![Screenshot 2021-03-28 at 11 15 54](https://user-images.githubusercontent.com/19173947/112748139-34f39600-8fba-11eb-9cf6-881760fffef3.png)

Generated files with custom directory structure
<img width="965" alt="Screenshot 2021-03-28 at 11 15 28" src="https://user-images.githubusercontent.com/19173947/112748147-4046c180-8fba-11eb-8462-1522947946f4.png">

**Sandbox URL**: N/A

**Merge deadline**: None

**Testing instructions**:

1. Start devstack
2. Change branch to `gabor/periodic-instructor-tasks`
3. Set `CELERYBEAT_SCHEDULER = 'djcelery.schedulers.DatabaseScheduler'` in LMS' `common.py` (for sake of simplicity)
4. Run `paver run_all_servers`
5. Go to LMS Django admin
6. Add a new periodic task [here](http://localhost:8000/admin/djcelery/periodictask) with the details below
7. Wait ~1 minute
8. Check `/tmp/edx-s3/` directory
9. Validate the report is generated

Task details:
Name: `test generate profile reports`
Task (registered): `periodic_instructor_reports.tasks.periodic_task_wrapper`
Interval -> Create a new interval for every minute
Arguments/Arguments: (note that the course ID will be different for you)
```
[
  ["course-v1:oc+oc+oc_1_1"],
  ["id", "username", "name", "email", "language", "location",
    "year_of_birth", "gender", "level_of_education", "mailing_address",
    "goals", "enrollment_mode", "verification_status"]
]
```

Arguments/Keyword arguments: (note that the user email should be your user's email)
```
{
  "task_name": "calculate_students_features_csv",
  "creator_email": "gabor@opencraft.com",
  "use_folders_by_date": true
}
```

**Author notes and concerns**:

1. Upstreaming probably will require us to allow task kwargs for every instructor task and not only for the two mentioned above.

**Reviewers**
